File: 力扣/每日一题/2022_04/04_22_旋转函数/main1.java
class Solution {
    // 规律 + DP
    public int maxRotateFunction(int[] nums) {
        // f(n) = f(n - 1) + sum - len * nums[len - n]
        int[] dp = new int[nums.length];
        int len = nums.length;
        // 初始化DP数组, 同时求数组和
        int sum = 0;
        for (int i = 0; i < len; i++) {
            dp[0] += i * nums[i];
            sum += nums[i];
        };
        // 递推, 并求最大值
        int max = dp[0];
        for (int i = 1; i < len; i++) {
            dp[i] = dp[i - 1] + sum - len * nums[len - i];
            if (dp[i] > max) max = dp[i];
        }
        return max;
    }
}

File: 力扣/每日一题/2022_06/06_07_爱吃香蕉的珂珂/main1.go
package main

// 二分搜索
func minEatingSpeed(piles []int, h int) int {
	l, r := 1, max(piles)

	// 判断以某个速度是否能在 h 小时内吃掉全部香蕉
	canEatAll := func(v int) bool {
		time := 0
		for _, n := range piles {
			// 时间 = 数量 / 速度
			if n%v == 0 {
				time += n / v
			} else {
				time += n/v + 1
			}
		}
		return time <= h
	}

	// 二分搜索
	for l <= r {
		m := l + (r-l)>>1
		if canEatAll(m) {
			r = m - 1
		} else {
			l = m + 1
		}
	}
	return l
}

func max(nums []int) int {
	res := nums[0]
	for _, v := range nums {
		if v > res {
			res = v
		}
	}
	return res
}
